bp_fe/src/include/bp_fe_pc_gen.vh

```diff
@@ -0,0 +1,208 @@
+ * Branch History Table (BHT).  If the BHT entry predicts taken, BP find the
+ * correpsonding entry in the Branch Target Buffer (BTB) as the next PC.  In the
+ * case of call-ret pairs, BP pushes the call target to Return Address Stack
+ * (RAS) and pops the address in the ret instruction.
+*/
+
+`ifndef BP_FE_PC_GEN_VH
+`define BP_FE_PC_GEN_VH
+
+`ifndef BSG_DEFINES_V
+`define BSG_DEFINES_V
+`include "bsg_defines.v"
+`endif
+
+`ifndef BP_COMMON_FE_BE_IF_VH
+`define BP_COMMON_FE_BE_IF_VH
+`include "bp_common_fe_be_if.vh"
+`endif
+
+// import bp_common_pkg::*;
+import pc_gen_pkg::*;
+
+typedef enum  {
+    e_state_reset               =0
+    ,e_state_first_reset_icache =1
+    ,e_state_first_start_icache =2
+    ,e_state_pc_redirect        =3
+    ,e_state_after_pc_redirect  =4
+    ,e_state_interrupt          =5
+    ,e_state_icache_miss        =6
+    ,e_state_after_icache_miss  =7
+    ,e_state_attaboy            =8
+    ,e_state_no_cmd             =9
+    ,e_state_not_ready          =10
+    ,e_state_always_not_ready   =11
+    ,e_state_invalid_cmd        =12
+} pc_gen_state_e;
+
+// this part is modified to accomodate the instr output
+`define declare_bp_fe_pc_gen_queue_s                    \
+    typedef struct packed {                             \
+        bp_fe_queue_type_e    msg_type;                 \
+        logic [`bp_fe_instr_scan_width-1:0] scan_instr; \
+        union packed {                                  \
+            bp_fe_fetch_s        fetch;                 \
+            bp_fe_exception_s    exception;             \
+        } msg;                                          \
+    } bp_fe_pc_gen_queue_s
+
+`define bp_fe_pc_gen_queue_width(vaddr_width_p,branch_metadata_fwd_width_p) \
+     (`bp_fe_queue_width(vaddr_width_p,branch_metadata_fwd_width_p)+`bp_fe_instr_scan_width)
+
+
+/*
+ * The pc_gen logic recieves the commands from the backend if there is any
+ * exceptions.  pc_gen inherets the interfaces from the frontend that uses the
+ * exception codes to notify if any exception happens.
+*/
+
+`define declare_bp_fe_pc_gen_cmd_s                                     \
+    typedef struct packed {                                            \
+        bp_fe_command_queue_opcodes_e          command_queue_opcodes;  \
+        union packed {                                                 \
+            bp_fe_cmd_pc_redirect_operands_s   pc_redirect_operands;   \
+            bp_fe_cmd_attaboy_s                attaboy;                \
+        } operands;                                                    \
+    } bp_fe_pc_gen_cmd_s
+
+
+`define bp_fe_pc_gen_cmd_width(vaddr_width_p,paddr_width_p,asid_width_p,branch_metadata_fwd_width_p) \
+    (`bp_fe_cmd_width(vaddr_width_p,paddr_width_p,asid_width_p,branch_metadata_fwd_width_p))
+
+
+/*
+ * bp_fe_pc_gen_s is provided to the backend, which consists of pc, BTB entry
+ * index, and ras address.
+*/
+`define bp_fe_pc_gen_fetch_s bp_fe_fetch_s
```

Comment:
Why redefine?

Reply:
Both redefinitions for bp_fe_pc_gen_fetch_s and bp_fe_pc_gen_exception_s are removed.